### PR TITLE
Fixes #1384. Fixes Scope.parseScopedKey

### LINKED ISSUE #1384: scoped key parsing
The regex used in Scope.parseScopedKey

```
  val ScopedKeyRegex = """((\w+)\/)?((\w+)\:)?([\w\-]+)""".r
```

does not parse scoped keys. \w+ is far too restrictive to match URIs, as the string is [defined to contain](http://www.scala-sbt.org/0.13.5/docs/Getting-Started/Scopes.html) - specifically "sbt displays (and parses) scoped keys like this" - and as sbt displays them.

```
scala> Scope parseScopedKey "{file:/mirror/r/java-stream-wrappers/}plot/runtime:classpathConfiguration"
scala.MatchError: {file:/mirror/r/java-stream-wrappers/}plot/runtime:classpathConfiguration (of class java.lang.String)
    at sbt.Scope$.parseScopedKey(Scope.scala:127)

scala> Scope parseScopedKey "plot/runtime:classpathConfiguration"
res1: (sbt.Scope, String) = (Scope(Select(LocalProject(plot)),Select(ConfigKey(runtime)),This,This),classpathConfiguration)
```

### PR DESCRIPTION
**Pull request by eed3si9n**
Fixes #1384.

`Scope.parseScopedKey` now supports full range of legal keys described in the documentation including {.} and other notation for `ProjectRef`, `BuildRef`, and `ThisBuild`.
Just in case any plugin is using `val ScopedKeyRegex = """((\w+)\/)?((\w+)\:)?([\w\-]+)""".r`, I'm leaving that old regex as is and added `ScopedKeyRegex2` in `private[sbt]`.

### CODE REVIEW

File: main/settings/src/main/scala/sbt/Scope.scala [hunk 1]
```diff
@@ -124,12 +124,32 @@ object Scope {
...
+          }
+          val configScope = configOpt map { case c if c != GlobalStr => Select(ConfigKey(c)) } getOrElse This
+          val inTaskScope = inTaskOpt map { t => Select(AttributeKey(t)) } getOrElse This
+          Scope(projScope, configScope, inTaskScope, This)
+      }
+      (scope, transformTaskName(key))
     }
+  // TODO: Change to ScopedKeyRegex2 on sbt 1.0
   val ScopedKeyRegex = """((\w+)\/)?((\w+)\:)?([\w\-]+)""".r
+  private[sbt] val ScopedKeyRegex2 = """([{](.*?)[}])?((\w*)\/)?(([\w\*]+)\:)?(([\w\-]+)\:\:)?([\w\-]+)""".r
```
Comment (jsuereth) [inline main/settings/src/main/scala/sbt/Scope.scala]: This looks like gobbledygook to me.  Any reason we don't just make it a format parser with less crazy?

I'm not 100% happy with the error messages you get if you f up the syntax already, but I'm not sure I'd be able to maintain this regex.

File: main/settings/src/main/scala/sbt/Scope.scala [hunk 2]
```diff
@@ -124,12 +124,32 @@ object Scope {
...
+            case _                        => This
+          }
+          val configScope = configOpt map { case c if c != GlobalStr => Select(ConfigKey(c)) } getOrElse This
+          val inTaskScope = inTaskOpt map { t => Select(AttributeKey(t)) } getOrElse This
+          Scope(projScope, configScope, inTaskScope, This)
+      }
+      (scope, transformTaskName(key))
     }
+  // TODO: Change to ScopedKeyRegex2 on sbt 1.0
   val ScopedKeyRegex = """((\w+)\/)?((\w+)\:)?([\w\-]+)""".r
```
Comment (jsuereth) [inline main/settings/src/main/scala/sbt/Scope.scala:153]: Mark this deprecated.

File: main/settings/src/test/scala/ScopedKeySpec.scala [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+class ScopedKeySpec extends Specification {
+  def is = s2"""
+
+  This is a specification to check the scoped key parsing.
+
+  fullClasspath should
+    ${beParsedAs("fullClasspath", ThisScope, "fullClasspath")}
+
+  test:fullClasspath should
+    ${beParsedAs("test:fullClasspath", ThisScope in ConfigKey("test"), "fullClasspath")}
```
Comment (jsuereth) [inline main/settings/src/test/scala/ScopedKeySpec.scala:19]: Love these tests.

**Comment by jsuereth** [comment]
The idea LGTM, but I think it's time to drop the Regexes and only have 1 problem.

**Comment by eed3si9n** [comment]
Now with deprecation.

**Comment by jsuereth** [comment]
LGTM, can you rebase so we can see a green CI?

**Pull request merged**